perl/lib/YAMLScript.pm

```diff
@@ -4,45 +4,78 @@ use strict;
-our $libyamlscript_version = '0.1.34';
 
 use Alien::YAMLScript;
-use FFI::Platypus;
 use FFI::CheckLib ();
+use FFI::Platypus;
 use JSON ();
 
-sub new {
-    bless {}, shift;
-}
-
-unless ( Alien::YAMLScript->exact_version($libyamlscript_version) ) {
-    my $have = Alien::YAMLScript->version;
-    die "YAMLScript $VERSION requires Alien::YAMLScript libyamlscript version\n" .
-        "$libyamlscript_version, but you have $have";
-}
+# Alien::YAMLScript finds the proper libyamlscript version, but we need to be
+# using the proper version of Alien::YAMLScript:
+die "\$YAMLScript::VERSION ($YAMLScript::VERSION) and " .
+    "\$Alien::YAMLScript::VERSION($Alien::YAMLScript::VERSION) " .
+    "must be the same version"
+    unless $YAMLScript::VERSION eq $Alien::YAMLScript::VERSION;
 
+# Set up FFI functions:
 my $ffi = FFI::Platypus->new(
     api => 2,
     lib => [ Alien::YAMLScript->dynamic_libs ],
 );
 
-$ffi->function( graal_create_isolate => [qw( opaque opaque* opaque* )] => 'int' )
-    ->( undef, \my $isolate, \my $thread)
-    and die 'Could not setup evaluation';
+my $graal_create_isolate = $ffi->function(
+    graal_create_isolate =>
+        ['opaque', 'opaque*', 'opaque*'] => 'int',
+);
 
-$ffi->attach(
-    [ load_ys_to_json => 'load' ] => [qw( sint64 string )] => 'string' => sub {
-        my ( $xsub, $self, $ys ) = @_;
-        my $resp = JSON::decode_json( $xsub->( $thread, $ys ) );
+my $graal_tear_down_isolate = $ffi->function(
+    graal_tear_down_isolate =>
+        ['opaque*'] => 'int',
+);
 
-        return $resp->{data} if $resp->{data};
+# YAMLScript object constuctor. Creates and saves a graal isolate thread:
+sub new {
+    my ($class, $config) = (@_, {});
+    my ($isolatethread);
+    $graal_create_isolate->(undef, undef, \$isolatethread) == 0
+        or die 'Failed to create graal isolate';
+    bless {
+        isolatethread => \$isolatethread,
+    }, $class;
+}
+
+# YAMLScript->load method.
+# Load a YAMLScript code string to produce a Perl object:
+sub _load {
+    my ($xsub, $self, $ys) = @_;
+    $self->{error} = undef;
+
+    my $resp = JSON::decode_json(
+        $xsub->(${$self->{isolatethread}}, $ys)
+    );
 
-        if (my $error = $resp->{error}{cause}) {
-            die "libyamlscript: $error";
-        }
+    return $resp->{data} if exists $resp->{data};
 
-        die "YAMLScript unkown error";
+    if ($self->{error} = $resp->{error}) {
+        die "libyamlscript: $self->{error}{cause}";
     }
+
+    die "Unexpected response from 'libyamlscript'";
+}
+
+# Attach the YAMLScript->load method to the libyamlscript load_ys_to_json
+# function:
+$ffi->attach(
+    [load_ys_to_json => 'load'] =>
+        ['sint64', 'string'] => 'string' =>
+        \&_load,
 );
 
+# Tear down the graal isolate when the YAMLScript object goes out of scope:
+sub DESTROY {
+    my ($self) = @_;
+    $graal_tear_down_isolate->(${$self->{isolatethread}}) == 0
+        or die "Failed to tear down graal isolate";
```

Comment:
Dying during destruction seems odd to me. If the teardown process fails, is there anything the user can do?

Reply:
Well, it did let me know that the tear down failed.
The raku binding has the same check but doesn't fail.
We need to sort out why the tear down fails.
I think it's a good check.

Reply:
I thought I had explained why the tear down fails. The change from `opaque*` to `opaque` should fix it

Reply:
What happens on a die during global destruction?